File: frontend/src/api/workouts.ts
import client from "./client";
import type { Workout, WorkoutDetail, WorkoutExercise } from "../types";

export const getWorkouts = () =>
  client.get<Workout[]>("/workouts").then((r) => r.data);

export const getWorkout = (id: number) =>
  client.get<WorkoutDetail>(`/workouts/${id}`).then((r) => r.data);

export const createWorkout = (data: {
  date: string;
  name?: string;
  notes?: string;
}) => client.post<Workout>("/workouts", data).then((r) => r.data);

export const updateWorkout = (
  id: number,
  data: { name?: string; date?: string; notes?: string }
) => client.put<Workout>(`/workouts/${id}`, data).then((r) => r.data);

export const deleteWorkout = (id: number) =>
  client.delete(`/workouts/${id}`);

export const addExerciseToWorkout = (
  workoutId: number,
  data: { exercise_id: number; order_index?: number; notes?: string }
) =>
  client
    .post<WorkoutExercise>(`/workouts/${workoutId}/exercises`, data)
    .then((r) => r.data);

export const removeExerciseFromWorkout = (
  workoutId: number,
  weId: number
) => client.delete(`/workouts/${workoutId}/exercises/${weId}`);

File: backend/app/routers/workouts.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.database import get_db
from app.schemas.workout import (
    WorkoutCreate, WorkoutUpdate, WorkoutRead, WorkoutDetail,
    WorkoutExerciseCreate, WorkoutExerciseRead,
)
from app.crud import workout as crud

router = APIRouter(prefix="/workouts", tags=["workouts"])


@router.get("", response_model=list[WorkoutRead])
def list_workouts(db: Session = Depends(get_db)):
    return crud.get_all(db)


@router.get("/{workout_id}", response_model=WorkoutDetail)
def get_workout(workout_id: int, db: Session = Depends(get_db)):
    workout = crud.get_detail(db, workout_id)
    if not workout:
        raise HTTPException(status_code=404, detail="Workout not found")
    return workout


@router.post("", response_model=WorkoutRead, status_code=201)
def create_workout(data: WorkoutCreate, db: Session = Depends(get_db)):
    return crud.create(db, data)


@router.put("/{workout_id}", response_model=WorkoutRead)
def update_workout(workout_id: int, data: WorkoutUpdate, db: Session = Depends(get_db)):
    workout = crud.get_by_id(db, workout_id)
    if not workout:
        raise HTTPException(status_code=404, detail="Workout not found")
    return crud.update(db, workout, data)


@router.delete("/{workout_id}", status_code=204)
def delete_workout(workout_id: int, db: Session = Depends(get_db)):
    workout = crud.get_by_id(db, workout_id)
    if not workout:
        raise HTTPException(status_code=404, detail="Workout not found")
    crud.delete(db, workout)


@router.post("/{workout_id}/exercises", response_model=WorkoutExerciseRead, status_code=201)
def add_exercise_to_workout(
    workout_id: int, data: WorkoutExerciseCreate, db: Session = Depends(get_db)
):
    workout = crud.get_by_id(db, workout_id)
    if not workout:
        raise HTTPException(status_code=404, detail="Workout not found")
    return crud.add_exercise(db, workout_id, data)


@router.delete("/{workout_id}/exercises/{we_id}", status_code=204)
def remove_exercise_from_workout(
    workout_id: int, we_id: int, db: Session = Depends(get_db)
):
    we = crud.get_workout_exercise(db, we_id)
    if not we or we.workout_id != workout_id:
        raise HTTPException(status_code=404, detail="WorkoutExercise not found")
    crud.remove_exercise(db, we)
